include/deal.II/base/mpi.templates.h

```diff
@@ -457,80 +457,100 @@ namespace Utilities
-                           result.end());
+              const unsigned int rank_recv =
+                (2 * stride) * (rank / (2 * stride));
+              const unsigned int rank_send = rank_recv + stride;
+
+              if (rank_send >= nproc) // nothing to do
+                continue;
+
+              if (rank_recv == rank) // process receives data
+                {
+                  MPI_Status status;
+                  const auto ierr_1 = MPI_Probe(rank_send,
+                                                internal::Tags::compute_union,
+                                                comm,
+                                                &status);
+                  AssertThrowMPI(ierr_1);
+
+                  int        amount;
+                  const auto ierr_2 = MPI_Get_count(&status, MPI_CHAR, &amount);
+                  AssertThrowMPI(ierr_2);
+
+                  std::vector<char> temp(amount);
+
+                  const auto ierr_3 = MPI_Recv(temp.data(),
+                                               amount,
+                                               MPI_CHAR,
+                                               status.MPI_SOURCE,
+                                               status.MPI_TAG,
+                                               comm,
+                                               MPI_STATUS_IGNORE);
+                  AssertThrowMPI(ierr_3);
+
+                  result = binary_op(result, Utilities::unpack<T>(temp, false));
+                }
+              else if (rank_send == rank) // process sends data
+                {
+                  std::vector<char> temp = Utilities::pack(result, false);
+
+                  const auto ierr = MPI_Send(temp.data(),
+                                             temp.size(),
+                                             MPI_CHAR,
+                                             rank_recv,
+                                             internal::Tags::compute_union,
+                                             comm);
+                  AssertThrowMPI(ierr);
+                }
             }
-          else if (rank_send == rank) // process sends data
-            {
-              const auto ierr = MPI_Send(result.data(),
-                                         result.size(),
-                                         internal::mpi_type_id(result.data()),
-                                         rank_recv,
-                                         internal::Tags::compute_union,
-                                         comm);
-              AssertThrowMPI(ierr);
-            }
-        }
 
-      // 2) broadcast result
-      int size = result.size();
-      MPI_Bcast(&size, 1, MPI_INT, 0, comm);
-      result.resize(size);
-      MPI_Bcast(
-        result.data(), size, internal::mpi_type_id(vec.data()), 0, comm);
+          // 2) broadcast result
+          std::vector<char> temp = Utilities::pack(result, false);
+          int               size = temp.size();
```

Comment:
```suggestion
          const int               size = temp.size();
```

Reply:
Could this possibly overflow? Maybe use the correct data type for the second argument of `MPI_Bcast`.

Reply:
I have changed the type to `unsigned int`. The variable cannot be made const due to the bcast.